---
title: "HiC-matrix_for_virt4c_norm"
author: "Jonas Ungerb√§ck"
date: "11/1/2019"
output: html_document
---

The purpose of this code it to read in and normalize FLWT and FLEKO PLAC-seq fata with the hicnorm package.  The data should be exported so it can be to to a dense matrix with HiC-Pro and then to a BUTLR object for virtual4c. Additionally, it should be exported so it can be overlapped with binding sites or other features in the bed-format.



```{r HiCcompare_normalization, echo=FALSE}
#multiHiCcompare is only recommended when you have multiple replicates so the developers recommend HiCcompare
library(HiCcompare)
library(preprocessCore)
setwd("/mnt/data/common/jonas/placseq/HiC-Pro-valid_interaction_files/multiHiCcompare_norm")
mat_WT <- read.table("/mnt/data/common/jonas/placseq/191009_21_174_175_H3K4me3_PLAC_FLWT_hicpro/hic_results/matrix/21_174_175_fastq_trimmed/raw/5000/21_174_175_fastq_trimmed_5000.matrix")
bed_WT <- read.table("/mnt/data/common/jonas/placseq/191009_21_174_175_H3K4me3_PLAC_FLWT_hicpro/hic_results/matrix/21_174_175_fastq_trimmed/raw/5000/21_174_175_fastq_trimmed_5000_abs.bed")
mat_EKO <- read.table("/mnt/data/common/jonas/placseq/191009_13_16_176_177_H3K4me3_PLAC_FLEKO_hicpro/hic_results/matrix/13_16_176_177_fastq_trimmed/raw/5000/13_16_176_177_fastq_trimmed_5000.matrix")
bed_EKO <- read.table("/mnt/data/common/jonas/placseq/191009_13_16_176_177_H3K4me3_PLAC_FLEKO_hicpro/hic_results/matrix/13_16_176_177_fastq_trimmed/raw/5000/13_16_176_177_fastq_trimmed_5000_abs.bed")

dat_WT <- HiCcompare::hicpro2bedpe(mat_WT, bed_WT)
dat_EKO <- HiCcompare::hicpro2bedpe(mat_EKO, bed_EKO)
#Bring out the cis interactions only
dat_WT <- dat_WT$cis
dat_EKO <- dat_EKO$cis

#If I want to use multiHiC-compare I have to pick dat_WT$cis and EKO$cis part chromosome by chromosome and then put these together into a df that I then rbind to one large object and feed into make_hicexp()
hic.list <- mapply(create.hic.table, dat_WT, dat_EKO, SIMPLIFY = FALSE, scale = FALSE)
hic.list <- total_sum(hic.list)
hic.list <- hic_loess(hic.list, parallel = TRUE)
# Merge the lists to a df
hic.list <- do.call(rbind, hic.list)
cor(hic.list$adj.IF1,hic.list$adj.IF2) # 0.6986924

hic_list4virt4c <- hic.list[,c(1:6,11,12)]

#Now hic.lists should be prepared for virtual 4c analysis as above
#I now have a df with adjusted IFs.
#Pull out the coordinates and adjusted IF, match it to the bin_keys and sort in original order.
hic_list4virt4c$bin1 <- paste(hic_list4virt4c$chr1,hic_list4virt4c$start1,hic_list4virt4c$end1,sep = "_")
hic_list4virt4c$bin2 <- paste(hic_list4virt4c$chr2,hic_list4virt4c$start2,hic_list4virt4c$end2,sep = "_")
#Replace the bins with the bin_keys:
gc_trans_key <- bed_WT
gc_trans_key$bin_key <- paste(gc_trans_key$V1, gc_trans_key$V2, gc_trans_key$V3, sep = "_")

#Replace the coordinates with the bin_key.
hic_list4virt4c$bin1 <- gc_trans_key$V4[match(hic_list4virt4c$bin1, gc_trans_key$bin_key)]
hic_list4virt4c$bin2 <- gc_trans_key$V4[match(hic_list4virt4c$bin2, gc_trans_key$bin_key)]
 
hic_list4virt4c <- hic_list4virt4c[,c(9,10,7,8)]
#Can I make a virtual 4c file from this or do I need to create the original length with zeros.
#Try with the wt first:
#It did work but unlike the the quantile norm from the ice norm above the base line between the wt and eko is different. Try a quant norm on this data and fo again:
x <- normalize.quantiles(as.matrix(hic_list4virt4c[,3:4]))
hic_list4virt4c <- cbind(hic_list4virt4c, as.data.frame(x))
#Changes correlation to 0.90.



write.table(hic_list4virt4c[,c(1,2,5)], "WT_hiccompare_norm_for_BUTR.matrix", sep = "\t", quote = F, row.names = F, col.names = F)

write.table(hic_list4virt4c[,c(1,2,6)], "EKO_hiccompare_norm_for_BUTR.matrix", sep = "\t", quote = F, row.names = F, col.names = F)

#In bash: /mnt/data/bioinfo_tools_and_refs/bioinfo_tools/HiC-Pro_2.11.1/bin/utils/sparseToDense.py -b /mnt/data/common/jonas/placseq/191009_21_174_175_H3K4me3_PLAC_FLWT_hicpro/hic_results/matrix/21_174_175_fastq_trimmed/raw/5000/21_174_175_fastq_trimmed_5000_abs.bed WT_hiccompare_norm_for_BUTR.matrix --perchr
#perl /mnt/data/bioinfo_tools_and_refs/bioinfo_tools/BUTLRTools/matrixToButlr.pl -a mm10 -g /mnt/data/common/jonas/placseq/mm10_chrNameLength.txt -m /mnt/data/common/jonas/placseq/HiC-Pro-valid_interaction_files/multiHiCcompare_norm/WT_chrom_matrix_file_hicnorm -r 5000 -o WT_chrom_matrix_file_hicnorm.butlr

#/mnt/data/bioinfo_tools_and_refs/bioinfo_tools/HiC-Pro_2.11.1/bin/utils/sparseToDense.py -b /mnt/data/common/jonas/placseq/191009_13_16_176_177_H3K4me3_PLAC_FLEKO_hicpro/hic_results/matrix/13_16_176_177_fastq_trimmed/raw/5000/13_16_176_177_fastq_trimmed_5000_abs.bed EKO_hiccompare_norm_for_BUTR.matrix --perchr
#perl /mnt/data/bioinfo_tools_and_refs/bioinfo_tools/BUTLRTools/matrixToButlr.pl -a mm10 -g /mnt/data/common/jonas/placseq/mm10_chrNameLength.txt -m /mnt/data/common/jonas/placseq/HiC-Pro-valid_interaction_files/multiHiCcompare_norm/EKO_chrom_matrix_file_hicnorm -r 5000 -o EKO_chrom_matrix_file_hicnorm.butlr
save.image("multiHiC_compare_norm.RData")

#Plot the densities to see that the lines do not overlap.
plot(density(hic_list4virt4c$adj.IF1), xlim=c(0,20))
lines(density(hic_list4virt4c$adj.IF1), col=c("red"))

#Diff compare
hic.list2 <- hic_compare(hic.list, A.min = 1, adjust.dist = TRUE, p.method = 'fdr', Plot = FALSE)
hic.list2 <- do.call(rbind, hic.list2)
hic.list2 <- hic.list2[hic.list2$p.adj<0.05,]

```

```{r man_virt4c, echo=FALSE}
install.packages("DescTools")
library(DescTools)
library(pheatmap)
library(RColorBrewer)
#How can I compare the signal over different regions? The way I see it we can take two different approaches. 1) We calculate total signal (maybe AUC) between WT and EKO for a given promoter and everything that connects to it. 2) One specify a set of bins and calculate the signal there. I will try to functionalize this but first I must write the code.

#Figure out what bin each promoter is in and assign it a key.
#Bring out mm10 tss: annotatePeaks.pl tss mm10 > mm10_tss.txt. Can I use bedtools to merge this with the bin bed file
# pos2bed mm10_tss.txt > mm10_tss.bed # bedtools intersect -wa -a 21_174_175_fastq_trimmed_5000_abs.bed -b mm10_tss_size0.bed > bins_5000_with_promoters.bed  #bedtools intersect -wb -a 21_174_175_fastq_trimmed_5000_abs.bed -b mm10_tss_size0.bed > promoters_in_bins_5000.bed

#########################################################################################################################
###################################       PREPROCESSING STARTS       #################################################### 
#########################################################################################################################
bins_5000_with_promoters <-read.delim("multiHiCcompare_norm/bins_5000_with_promoters.bed", header = F, stringsAsFactors = F)
promoters_in_bins_5000 <- read.delim("multiHiCcompare_norm/promoters_in_bins_5000.bed", header = F, stringsAsFactors = F)

#From promoters_in_bins bring out key and corresponding key. Looks like they are sorted in the same order
promoters_in_bins_5000 <- promoters_in_bins_5000[,c(4,8)]
bins_5000_with_promoters <- merge(bins_5000_with_promoters, promoters_in_bins_5000, by = "V4", all.x = T)
#Several promoters cam be in one bin so
bins_5000_with_promoters$unique_entries <- paste(bins_5000_with_promoters$V4,bins_5000_with_promoters$V8, sep = "_")
bins_5000_with_promoters <- bins_5000_with_promoters[!duplicated(bins_5000_with_promoters$unique_entries),]
bins_5000_with_promoters <- bins_5000_with_promoters[,-6]
#Now bring in more annotations to this list:
mm10_tss <- read.delim("multiHiCcompare_norm/mm10_tss_size0.txt", stringsAsFactors = F)
colnames(mm10_tss)[1]<- "PeakID"
mm10_tss <- mm10_tss[,c(1,12,14:16)]

promoter_bin_key <- merge(bins_5000_with_promoters, mm10_tss, by.x = "V8", by.y = "PeakID", all.x = T)
colnames(promoter_bin_key)[1:5] <- c("PromoterID","bin_key", "chr", "start", "end") 
#Reorder after bin_key
promoter_bin_key <- promoter_bin_key[order(promoter_bin_key$bin_key),]

colnames(hic_list4virt4c)[5:6]<-c("quant.IF1","quant.IF2")
# Filter out lines that has the same bins interacting.
hic_list4virt4c_no_same <- hic_list4virt4c[-which(hic_list4virt4c$bin1 == hic_list4virt4c$bin2),]

#########################################################################################################################
###################################       PREPROCESSING ENDS      #################################################### 
#########################################################################################################################
#I should be able to use this key to match a gene to the total score that is connected to one bin.
#try an exampel with Pax5

#The genebinIntScore calculates the area-under-curve (AUC) for HiCcompare-normalized FLWT and FLEKO interactions connected to a given gene.

genebinIntScore <- function(df = hic_list4virt4c_no_same, gene_name){
  library(DescTools)
  #Create a list that of all the Pax5 connected bins that I can later loop over
  bin_key <- lapply(unique(promoter_bin_key[ which(promoter_bin_key$Gene.Name == gene_name), ]$bin_key) , as.data.frame )  

  #Create the AUC for all the bins connecting to bin_key[[i]
  mat <- matrix(, nrow = length(bin_key), ncol = 2)
  name_vector <- vector(length = length(bin_key) )
  #Loop over all the bins:
  for (i in 1:length(bin_key)) {
    bin_nr <- as.numeric( bin_key[[i]] )
    print(bin_nr)
    #Take out the rows matching these bins from hic_list4virt4c_no_same
    int_bins <- subset(hic_list4virt4c_no_same, hic_list4virt4c_no_same$bin1 == bin_nr | hic_list4virt4c_no_same$bin2 == bin_nr)
    #Add a counter if ot doesn't exist.:
    if (!"counter" %in% colnames(int_bins)){
    int_bins$counter <- seq(1,length(int_bins[,1]))
    }

    mat[i,1] <- as.numeric( AUC(x = as.vector(int_bins$quant.IF1), y = as.vector(int_bins$counter)) )
    mat[i,2] <- as.numeric( AUC(x = as.vector(int_bins$quant.IF2), y = as.vector(int_bins$counter)) )
    name_vector[i]  <- paste(gene_name, bin_nr, sep = "_")
 
    #Loop ends here
    print(i)
  }
results_AUC <- as.data.frame(mat)
colnames(results_AUC) <- c("AUC_FLWT","AUC_FLEKO")
rownames(results_AUC) <- name_vector
return(results_AUC)

}

genes_test <- c("Gata3", "Ebf1", "Pax5","Hprt","Notch1","Cd79b","Igll1","Cd19","Tcf7", "Actb","Gapdh", "Flt3","Cd5", "Bcl11b", "Runx1", "Ikzf1", "Ikzf2", "Ikzf3", "Myc")

#Match this in to
idToGene <- promoter_bin_key[,c(6,9)]

#Real data!!!! Exciting!
genes_w_WT_and_EKO_acc_ATAC <- read.delim("multiHiCcompare_norm/gene_list_analysis/Genes_w_WT_and_EKO_acc_ATAC.csv", stringsAsFactors = F)
colnames(genes_w_WT_and_EKO_acc_ATAC)[1] <- "Entrez.ID"
genes_w_WT_and_EKO_acc_ATAC <- merge(genes_w_WT_and_EKO_acc_ATAC,idToGene, by = "Entrez.ID", all.x = T)
genes_w_WT_and_EKO_acc_ATAC <- genes_w_WT_and_EKO_acc_ATAC[!duplicated(genes_w_WT_and_EKO_acc_ATAC$Gene.Name),]

genes_w_WT_only_acc_ATAC <- read.delim("multiHiCcompare_norm/gene_list_analysis/Genes_w_WT_only_acc_ATAC.csv", stringsAsFactors = F)
colnames(genes_w_WT_only_acc_ATAC)[1] <- "Entrez.ID"
genes_w_WT_only_acc_ATAC <- merge(genes_w_WT_only_acc_ATAC,idToGene, by = "Entrez.ID", all.x = T)
genes_w_WT_only_acc_ATAC <- genes_w_WT_only_acc_ATAC[!duplicated(genes_w_WT_only_acc_ATAC$Gene.Name),]

atac_access_mixed <- read.delim("multiHiCcompare_norm/gene_list_analysis/Genes_w_WT_and_EKO_acc_ATAC_Genes_w_WT_only_acc_ATAC.csv", stringsAsFactors = F)
colnames(atac_access_mixed)[1] <- "Entrez.ID"
atac_access_mixed <- merge(atac_access_mixed,idToGene, by = "Entrez.ID", all.x = T)
atac_access_mixed <- atac_access_mixed[!duplicated(atac_access_mixed$Gene.Name),]


#Turn to vectors_
genes_w_WT_and_EKO_acc_ATAC_vector <- as.vector(genes_w_WT_and_EKO_acc_ATAC['Gene.Name'])
genes_w_WT_only_acc_ATAC_vector <- as.vector(genes_w_WT_only_acc_ATAC['Gene.Name'])
atac_access_mixed_vector <- as.vector(atac_access_mixed['Gene.Name'])

#Turn the dfs and vectors above to a list for a cleaner workspace.
atac_gen_cat_list <- list(genes_w_WT_and_EKO_acc_ATAC = genes_w_WT_and_EKO_acc_ATAC, genes_w_WT_only_acc_ATAC = genes_w_WT_only_acc_ATAC, atac_access_mixed = atac_access_mixed, genes_w_WT_and_EKO_acc_ATAC_vector = genes_w_WT_and_EKO_acc_ATAC_vector, genes_w_WT_only_acc_ATAC_vector = genes_w_WT_only_acc_ATAC_vector, atac_access_mixed_vector = atac_access_mixed_vector)

genes <- j
#Create a loop over a vector and put the results in each entry
gene_virt4c_score <- vector(mode = "list", length = 0)
for (j in genes_test) {
  print(j)
  gene_virt4c_score[[j]] <- genebinIntScore(df = hic_list4virt4c, gene_name = j)
}

df <- do.call(rbind, lapply(gene_virt4c_score, data.frame, stringsAsFactors=FALSE))
df$WT <- df$AUC_FLWT/df$AUC_FLWT
df$EKO <- df$AUC_FLEKO/df$AUC_FLWT

#Plot the test genes:
breakList <- seq(-2, 2, by = 0.1)
pheatmap(mat = log2(df[,c(3,4)]), cluster_rows = FALSE, cluster_cols = FALSE, show_rownames = FALSE,  scale = "none" ,  color = colorRampPalette(rev(brewer.pal(n = 11, name = "RdYlBu")))(length(breakList)), breaks = breakList  )

df_genes_atac_access_mixed_vector <- df

#Give each df a unique cat variable.
df_genes_atac_access_mixed_vector$category <-"df_genes_atac_access_mixed"
df_genes_w_WT_and_EKO_acc_ATAC_vector$category  <- "df_genes_w_WT_and_EKO_acc_ATAC"
df_genes_w_WT_only_acc_ATAC$category <- "df_genes_w_WT_only_acc_ATAC"

#Order each list on EKO value and rbind them.
df_genes_atac_access_mixed_vector <- df_genes_atac_access_mixed_vector[order(df_genes_atac_access_mixed_vector$EKO),]
df_genes_w_WT_and_EKO_acc_ATAC_vector <- df_genes_w_WT_and_EKO_acc_ATAC_vector[order(df_genes_w_WT_and_EKO_acc_ATAC_vector$EKO),]
df_genes_w_WT_only_acc_ATAC <- df_genes_w_WT_only_acc_ATAC[order(df_genes_w_WT_only_acc_ATAC$EKO),]

df_genes_atac_cats <- rbind(df_genes_w_WT_only_acc_ATAC, df_genes_atac_access_mixed_vector,df_genes_w_WT_and_EKO_acc_ATAC_vector)

breakList <- seq(-3, 3, by = 0.1)
ann_cols <-as.data.frame(df_genes_atac_cats[,c(5,5)])
ann_cols$category <- as.factor(ann_cols$category)
ann_cols$category.1 <- as.factor(ann_cols$category.1)

rownames(ann_cols)<-rownames(df_genes_atac_cats)

pheatmap(mat = log2(df_genes_atac_cats[,c(4,4)]), cluster_rows = FALSE, cluster_cols = FALSE, show_rownames = FALSE,  scale = "none" ,  color = colorRampPalette(rev(brewer.pal(n = 11, name = "RdYlBu")))(length(breakList)), breaks = breakList , annotation_row = ann_cols[,c(1,1)] )

#Ok, now we have the heatmap. Split this in up and down in EKO and make two box_plots-
df_genes_atac_cats_up_EKO <- subset(df_genes_atac_cats,df_genes_atac_cats$EKO > 1)
df_genes_atac_cats_down_EKO <- subset(df_genes_atac_cats,df_genes_atac_cats$EKO < 1)

df_genes_atac_cats_up_EKO <- df_genes_atac_cats_up_EKO[,c(1,2,5)]
df_genes_atac_cats_down_EKO <- df_genes_atac_cats_down_EKO[,c(1,2,5)]

df_genes_atac_cats_up_EKO <- melt(data = df_genes_atac_cats_up_EKO, id.vars = c("category"), measure.vars = c("AUC_FLWT","AUC_FLEKO") )
df_genes_atac_cats_up_EKO$category <- factor(df_genes_atac_cats_up_EKO$category, levels = c("df_genes_w_WT_only_acc_ATAC", "df_genes_atac_access_mixed", "df_genes_w_WT_and_EKO_acc_ATAC"), )

ggplot() + geom_violin(data = df_genes_atac_cats_up_EKO,aes(x = category, y = log2(value), fill = variable),  draw_quantiles = c(0.25, 0.5, 0.75), scale = "count" ) + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + theme_bw()
ggsave("genes_atac_cats_up_EKO.pdf",device = "pdf", dpi = 300,width = 10, height = 8, units = c("cm"))

df_genes_atac_cats_down_EKO <- melt(data = df_genes_atac_cats_down_EKO, id.vars = c("category"), measure.vars = c("AUC_FLWT","AUC_FLEKO") )
df_genes_atac_cats_down_EKO$category <- factor(df_genes_atac_cats_down_EKO$category, levels = c("df_genes_w_WT_only_acc_ATAC", "df_genes_atac_access_mixed", "df_genes_w_WT_and_EKO_acc_ATAC"))

ggplot() + geom_violin(data = df_genes_atac_cats_down_EKO,aes(x = category, y = log2(value), fill = variable),  draw_quantiles = c(0.25, 0.5, 0.75), scale = "count" ) + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + theme_bw()
ggsave("genes_atac_cats_down_EKO.pdf",device = "pdf", dpi = 300,width = 10, height = 8, units = c("cm"))

#Statistical calulation on the groups in the plot
#Up_eko

wilcox.test(value ~variable, data = subset(df_genes_atac_cats_up_EKO, df_genes_atac_cats_up_EKO$category == "df_genes_w_WT_only_acc_ATAC")) # p=0.36
wilcox.test(value ~variable, data = subset(df_genes_atac_cats_up_EKO, df_genes_atac_cats_up_EKO$category == "df_genes_atac_access_mixed")) # p=0.34
wilcox.test(value ~variable, data = subset(df_genes_atac_cats_up_EKO, df_genes_atac_cats_up_EKO$category == "df_genes_w_WT_and_EKO_acc_ATAC")) # p=0.05


wilcox.test(value ~variable, data = subset(df_genes_atac_cats_down_EKO, df_genes_atac_cats_down_EKO$category == "df_genes_w_WT_only_acc_ATAC")) # p=0.02
wilcox.test(value ~variable, data = subset(df_genes_atac_cats_down_EKO, df_genes_atac_cats_down_EKO$category == "df_genes_atac_access_mixed")) # p=4.928e-12
wilcox.test(value ~variable, data = subset(df_genes_atac_cats_down_EKO, df_genes_atac_cats_down_EKO$category == "df_genes_w_WT_and_EKO_acc_ATAC")) # p=0.34


```

```{r bin_score_calculation_per_overlapping_bins, echo=FALSE}
source("/home/jonas/R_scripts/match_list_function.R")
library(dplyr)
library(reshape2)
library(ggplot2)
# Using bedtools intersect with the bin_key list and a peak-file, determine in what bin a peak exists. Use the 230238 Ebf1-peak list.
# pos2bed.pl Combined_230_238_a_Ebf1_M1_M44-top-set.mm10_annotated.txt > Combined_230_238_a_Ebf1_M1_M44-top-set.mm10_annotated.bed. # bedtools intersect -wa -a 21_174_175_fastq_trimmed_5000_abs.bed -b Combined_230_238_a_Ebf1_M1_M44-top-set.mm10_annotated.bed > bins_with_Ebf1_peaks.bed #bedtools intersect -wb -a 21_174_175_fastq_trimmed_5000_abs.bed -b Combined_230_238_a_Ebf1_M1_M44-top-set.mm10_annotated.bed > Ebf1_coordinates_in_bins.bed

bins_with_Ebf1_peaks <- read.delim("multiHiCcompare_norm/bins_with_Ebf1_peaks.bed", stringsAsFactors = F, header = F)
Ebf1_coordinates_in_bins <- read.delim("multiHiCcompare_norm/Ebf1_coordinates_in_bins.bed", stringsAsFactors = F, header = F)
bins_with_Ebf1_peaks <- cbind(bins_with_Ebf1_peaks, Ebf1_coordinates_in_bins[,8])
colnames(bins_with_Ebf1_peaks)[5]<-"PeakID"

#Can I use my match_list function?
bins_with_score <- hic_list4virt4c_no_same
#With my match_list function, deduce which interactions that have Ebf1 in one or two bins.
bins_with_score <- match_lists(bins_with_score, bins_with_score$bin1, bins_with_Ebf1_peaks$V4, 1, name_match_column = "bin1_Ebf1_bound")
bins_with_score <- match_lists(bins_with_score, bins_with_score$bin2, bins_with_Ebf1_peaks$V4, 1, name_match_column = "bin2_Ebf1_bound")
bins_with_score$Ebf1_bin_status <- bins_with_score$bin1_Ebf1_bound + bins_with_score$bin2_Ebf1_bound 
table(bins_with_score$Ebf1_bin_status)

bins_with_score %>% group_by(Ebf1_bin_status) %>% summarise(average=mean(quant.IF1))
bins_with_score %>% group_by(Ebf1_bin_status) %>% summarise(average=mean(quant.IF2))

#Plot on the spot:
test <- bins_with_score
test$bin_key <- paste(test$bin1,test$bin2, sep = "_")      
test <- test[,c(9,5,6)]
test <- melt(data = test, id.vars = c("Ebf1_bin_status"), measure.vars = c("quant.IF1","quant.IF2") )
test$Ebf1_bin_status <- as.factor(as.character(test$Ebf1_bin_status))
ggplot(data = test) + geom_boxplot(aes(x = Ebf1_bin_status, y = log10(value), fill = variable))


```

```{r HiCcompare_normalization_res_bin_size_data, echo=FALSE}
#multiHiCcompare is only recommended when you have multiple replicates so the developers recommend HiCcompare
library(multiHiCcompare)
library(HiCcompare)
library(preprocessCore)

#Turns out that I cannot do this. Data too sparse (too many 1s)

```

